File: backend/app/schemas/tour.py
from pydantic import BaseModel, Field
from typing import Optional, List, Any
from decimal import Decimal
from datetime import datetime
from enum import Enum


class Region(str, Enum):
    NORTH = "NORTH"
    CENTRAL = "CENTRAL"
    SOUTH = "SOUTH"
    INTERNATIONAL = "INTERNATIONAL"


class TourBase(BaseModel):
    name: str
    slug: str
    description: Optional[str] = None
    short_description: Optional[str] = None
    destination: str
    region: Optional[Region] = None
    duration: str
    price: Decimal
    discount_price: Optional[Decimal] = None
    max_participants: int = 20
    images: List[str] = []
    highlights: List[str] = []
    includes: List[str] = []
    excludes: List[str] = []
    schedule: Optional[List[dict]] = None
    departure_dates: List[str] = []
    is_featured: bool = False
    category: Optional[str] = None
    tags: List[str] = []


class TourCreate(TourBase):
    pass


class TourUpdate(BaseModel):
    name: Optional[str] = None
    slug: Optional[str] = None
    description: Optional[str] = None
    short_description: Optional[str] = None
    destination: Optional[str] = None
    region: Optional[Region] = None
    duration: Optional[str] = None
    price: Optional[Decimal] = None
    discount_price: Optional[Decimal] = None
    max_participants: Optional[int] = None
    images: Optional[List[str]] = None
    highlights: Optional[List[str]] = None
    includes: Optional[List[str]] = None
    excludes: Optional[List[str]] = None
    schedule: Optional[List[dict]] = None
    departure_dates: Optional[List[str]] = None
    is_featured: Optional[bool] = None
    is_active: Optional[bool] = None
    category: Optional[str] = None
    tags: Optional[List[str]] = None


class TourResponse(TourBase):
    id: str
    current_participants: int = 0
    rating: Decimal = Decimal("0")
    review_count: int = 0
    is_active: bool = True
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class TourListResponse(BaseModel):
    tours: List[TourResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class TourFilter(BaseModel):
    destination: Optional[str] = None
    region: Optional[Region] = None
    category: Optional[str] = None
    min_price: Optional[Decimal] = None
    max_price: Optional[Decimal] = None
    duration: Optional[str] = None
    is_featured: Optional[bool] = None
    search: Optional[str] = None

File: frontend/types/tour.ts
// ============================================================
// Tour Types — mirrors backend schemas/tour.py + prisma Tour model
// ============================================================

export type Region = "NORTH" | "CENTRAL" | "SOUTH" | "INTERNATIONAL";

export interface TourImage {
  url: string;
  alt?: string;
}

export interface TourHighlight {
  icon?: string;
  text: string;
}

export interface TourScheduleDay {
  day: number;
  title: string;
  activities: string[];
  meals?: string[];
  notes?: string;
}

export interface Tour {
  id: string;
  name: string;
  slug: string;
  description?: string;
  short_description?: string;
  destination: string;
  region?: Region;
  duration: string; // e.g. "3 ngày 2 đêm"
  price: number;
  discount_price?: number;
  max_participants: number;
  current_participants: number;
  images: string[] | TourImage[];
  highlights?: string[] | TourHighlight[];
  includes?: string[];
  excludes?: string[];
  schedule?: TourScheduleDay[];
  departure_dates?: string[];
  rating: number;
  review_count: number;
  is_featured: boolean;
  is_active: boolean;
  category?: string;
  tags?: string[];
  created_at: string;
  updated_at?: string;
}

export interface TourFilters {
  destination?: string;
  region?: Region;
  category?: string;
  min_price?: number;
  max_price?: number;
  search?: string;
  is_featured?: boolean;
  page?: number;
  page_size?: number;
}

export interface TourListResponse {
  tours: Tour[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface TourCreate {
  name: string;
  slug?: string;
  description?: string;
  short_description?: string;
  destination: string;
  region?: Region;
  duration: string;
  price: number;
  discount_price?: number;
  max_participants?: number;
  images?: string[];
  highlights?: string[];
  includes?: string[];
  excludes?: string[];
  schedule?: TourScheduleDay[];
  departure_dates?: string[];
  is_featured?: boolean;
  category?: string;
  tags?: string[];
}

export interface TourUpdate extends Partial<TourCreate> {
  is_active?: boolean;
}

export interface Review {
  id: string;
  user_id: string;
  tour_id: string;
  booking_id?: string;
  rating: number;
  comment?: string;
  is_verified: boolean;
  created_at: string;
  user?: {
    full_name: string;
    avatar_url?: string;
  };
}

export interface TourWithReviews extends Tour {
  reviews?: Review[];
}
